src/Microsoft.AspNetCore.Authentication/Events/BaseContext.cs

```diff
@@ -9,41 +9,70 @@ namespace Microsoft.AspNetCore.Authentication
     /// <summary>
     /// Base class used by other context classes.
     /// </summary>
-    public abstract class BaseContext
+    public abstract class BaseContext<TOptions> where TOptions : AuthenticationSchemeOptions
     {
+        private AuthenticationProperties _properties;
+
         /// <summary>
         /// Constructor.
         /// </summary>
-        /// <param name="context">The request context.</param>
-        protected BaseContext(HttpContext context)
+        /// <param name="context">The context.</param>
+        /// <param name="scheme">The authentication scheme.</param>
+        /// <param name="options">The authentication options associated with the scheme.</param>
+        protected BaseContext(HttpContext context, AuthenticationScheme scheme, TOptions options)
         {
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
+            if (scheme == null)
+            {
+                throw new ArgumentNullException(nameof(scheme));
+            }
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
 
             HttpContext = context;
+            Scheme = scheme;
+            Options = options;
         }
 
         /// <summary>
+        /// The authentication scheme.
+        /// </summary>
+        public AuthenticationScheme Scheme { get; }
+
+        /// <summary>
+        /// Gets the authentication options associated with the scheme.
+        /// </summary>
+        public TOptions Options { get; }
+
+        /// <summary>
         /// The context.
         /// </summary>
         public HttpContext HttpContext { get; }
 
         /// <summary>
         /// The request.
         /// </summary>
-        public HttpRequest Request
-        {
-            get { return HttpContext.Request; }
-        }
+        public HttpRequest Request => HttpContext.Request;
 
         /// <summary>
         /// The response.
         /// </summary>
-        public HttpResponse Response
+        public HttpResponse Response => HttpContext.Response;
+
+        /// <summary>
+        /// Gets or sets the <see cref="AuthenticationProperties"/>.
+        /// </summary>
+        public virtual AuthenticationProperties Properties
         {
-            get { return HttpContext.Response; }
+            get => _properties ?? (_properties = new AuthenticationProperties());
+            set => _properties = value;
         }
+
+        public EventResultState State { get; protected set; }
```

Comment:
I'll restate what I said before in different terms :)  Basically if you limit the control changes to classes under BaseControlContext/BaseJwtBearer/BaseOpenIdConnect, we can wait to rationalize those with the rest of the events.  That's what I was referring to with my 'bleed' remark, Cookies/OAuth/Twitter context's should not be affected since they currently do not use BaseControlContext at all.